internal/charmstore/debug.go

```diff
@@ -14,60 +14,304 @@ import (
 	"strings"
 	"time"
 
+	"github.com/juju/httprequest"
 	"github.com/juju/utils"
+	"github.com/juju/utils/debugstatus"
+	"github.com/julienschmidt/httprouter"
 	"gopkg.in/errgo.v1"
 	"gopkg.in/juju/charmrepo.v2-unstable/csclient/params"
 	"gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
 
+	"gopkg.in/juju/charmstore.v5-unstable/internal/mongodoc"
 	"gopkg.in/juju/charmstore.v5-unstable/internal/router"
 	appver "gopkg.in/juju/charmstore.v5-unstable/version"
 )
 
-// GET /debug/info .
-func serveDebugInfo(http.Header, *http.Request) (interface{}, error) {
-	return appver.VersionInfo, nil
+type debugParams struct {
+	ServerParams
+	pool            *Pool
+	loopbackHandler http.Handler
 }
 
-// GET /debug/check.
-func debugCheck(checks map[string]func() error) http.Handler {
-	return router.HandleJSON(func(http.Header, *http.Request) (interface{}, error) {
-		n := len(checks)
-		type result struct {
-			name string
-			err  error
+type debugHandler struct {
+	router    *httprouter.Router
+	params    debugParams
+	checks    map[string]func() error
+	openPaths map[string]bool
+}
+
+func newDebugHandler(p debugParams) *debugHandler {
+	hnd := debugHandler{
+		router: httprouter.New(),
+		params: p,
+		checks: map[string]func() error{
+			"mongodb":       checkDB(p.pool.db.Database),
+			"elasticsearch": checkES(p.pool.es),
+		},
+		openPaths: map[string]bool{
+			"/debug/check":  true,
+			"/debug/info":   true,
+			"/debug/status": true,
+		},
+	}
+	for _, h := range router.ErrorToResp.Handlers(hnd.handle) {
+		hnd.router.Handle(h.Method, h.Path, h.Handle)
+	}
+	return &hnd
+}
+
+func (h *debugHandler) handle(p httprequest.Params) (*debugReqHandler, error) {
+	if err := h.authorized(p.Request); err != nil {
+		return nil, err
+	}
+	rh := debugReqHandler{
+		Handler: debugstatus.Handler{
+			Version: debugstatus.Version(appver.VersionInfo),
+			// Authorization has alread been checked so there is no need to do it again.
+			CheckPprofAllowed: func(*http.Request) error { return nil },
+		},
+		h: h,
+	}
+	rh.Handler.Check = rh.check
+	return &rh, nil
+}
+
+func (h *debugHandler) authorized(r *http.Request) error {
+	if h.openPaths[r.URL.Path] {
+		return nil
+	}
+	// TODO support other login methods
+	u, p, err := utils.ParseBasicAuthHeader(r.Header)
+	if err != nil {
+		return errgo.WithCausef(err, params.ErrUnauthorized, "")
+	}
+	if u != h.params.AuthUsername || p != h.params.AuthPassword {
+		return errgo.WithCausef(nil, params.ErrUnauthorized, "username or password mismatch")
+	}
+	return nil
+}
+
+func (h *debugHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// The charmstore router will have stripped "/debug" off the
+	// front of the path, put it back.
```

Comment:
Might be good to state *why* we're putting it back.
i.e.
// ... put it back. We need it because the handlers that debugstatus
// defined have the /debug prefix.